pylint/testutils/functional/find_functional_tests.py

```diff
@@ -49,39 +51,89 @@ def get_functional_test_files_from_directory(
+        return files_without_leading_underscore
+
+    def walk(path: Path) -> Iterator[Path]:
+        violations: list[tuple[Path, int]] = []
+        violations_msgs: set[str] = set()
+        parent_dir_files = _get_files_from_dir(path, violations)
+        error_msg = (
+            "The following directory contains too many functional tests files:\n"
+        )
+        for _file_or_dir in parent_dir_files:
+            if _file_or_dir.is_dir():
+                _files = _get_files_from_dir(_file_or_dir, violations)
+                yield _file_or_dir.resolve()
+                try:
+                    yield from walk(_file_or_dir)
+                except AssertionError as e:
+                    violations_msgs.add(str(e).replace(error_msg, ""))
+            else:
+                yield _file_or_dir.resolve()
+        if violations:
+            _msg = error_msg
+            for offending_file, number in violations:
+                _msg += f"- {offending_file}: {number} when the max is {max_file_per_directory}\n"
+            for error_msg in violations_msgs:
+                _msg += error_msg
+            raise AssertionError(_msg)
+
     # Collect all sub-directories and files in directory
-    for file_or_dir in directory.iterdir():
-        if file_or_dir.is_file():
-            if file_or_dir.suffix == ".py" and not file_or_dir.stem.startswith("_"):
-                files.add(file_or_dir)
-        elif file_or_dir.is_dir():
+    for file_or_dir in walk(directory):
+        if file_or_dir.stem.startswith("_"):
+            continue
+        if file_or_dir.is_dir():
             dirs.add(file_or_dir)
-            _check_functional_tests_structure(file_or_dir)
-
-    assert len(files) <= REASONABLY_DISPLAYABLE_VERTICALLY, (
-        f"{directory} contains too many functional tests files "
-        + f"({len(files)} > {REASONABLY_DISPLAYABLE_VERTICALLY})."
-    )
+        elif file_or_dir.suffix == ".py":
+            files.add(file_or_dir)
 
+    directory_does_not_exists: list[tuple[Path, Path]] = []
+    misplaced_file: list[Path] = []
     for file in files:
         possible_dir = file.parent / file.stem.split("_")[0]
-        assert not possible_dir.exists(), f"{file} should go in {possible_dir}."
-
+        if possible_dir.exists():
+            directory_does_not_exists.append((file, possible_dir))
         # Exclude some directories as they follow a different structure
         if (
             not len(file.parent.stem) == 1  # First letter sub-directories
             and file.parent.stem not in IGNORED_PARENT_DIRS
             and file.parent.parent.stem not in IGNORED_PARENT_PARENT_DIRS
         ):
-            assert file.stem.startswith(
-                file.parent.stem
-            ), f"{file} should not go in {file.parent}"
+            if not file.stem.startswith(file.parent.stem):
+                misplaced_file.append(file)
+    if directory_does_not_exists or misplaced_file:
```

Comment:
Please add some new lines in between 103-139. These blocks seem to served different purposes (like the comments indicate) so imo they should be separated

Reply:
2430c6a5e